src/forbidden_tree.jl

```diff
@@ -0,0 +1,188 @@
+	tree::AbstractConstraintMatchNode
+end
+
+
+# Matching RuleNode with ConstraintMatchNode
+"""
+Tries to match RuleNode `rn` with ConstraintMatchNode `cmn` and fill in 
+the domain of the hole at `hole_location`. 
+Returns if match is successful:
+
+- A dictionary with variable assignments
+- The id for the node which fills the hole
+A variable is represented by the index of its rulenode.
+Returns nothing if the match is unsuccessful.
+"""
+function _match_expr_with_hole(
+    rn::RuleNode, 
+    cmn::ConstraintMatchNode, 
+    hole_location::Vector{Int}
+)::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    hole_location == [] && throw(ArgumentError("The hole location doesn't point to a hole!"))
+
+    if rn.ind ≠ cmn.rule_ind || length(rn.children) ≠ length(cmn.children)
+        return nothing
+    else
+        vars = Dict()
+        forbidden_domain = nothing
+        for (i, rnᵢ, cmnᵢ) ∈ zip(1:length(rn.children), rn.children, cmn.children)
+            if i == hole_location[1]
+                # Match with finding domain of hole
+                forbidden_domain, varsᵢ = _match_expr_with_hole(rnᵢ, cmnᵢ, hole_location[begin+1:end])
+            else
+                # Regular match
+                varsᵢ = _match_expr(rnᵢ, cmnᵢ)
+            end
+
+            if varsᵢ ≡ nothing
+                # unsuccessful match
+                return nothing
+            end
+            # Check if another argument already assigned the same variables
+            for (k, v) ∈ varsᵢ
+                if k ∈ keys(vars) && v ≠ vars[k]
+                    return nothing
+                end
+                vars[k] = v
+            end
+        end
+        return forbidden_domain, vars
+    end
+end
+
+# Matching RuleNode with ConstraintMatchVar
+function _match_expr_with_hole(
+    ::RuleNode, 
+    ::ConstraintMatchVar, 
+    ::Vector{Int}
+)::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    return nothing
+end
+
+# Matching Hole with ConstraintMatchNode
+function _match_expr_with_hole(
+    ::Hole, 
+    cmn::ConstraintMatchNode, 
+    hole_location::Vector{Int}
+ )::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    if hole_location == [] && cmn.children == []
+        return cmn.rule_ind, Dict()
+    end
+    return nothing
+end
+
+# Matching Hole with ConstraintMatchVar
+function _match_expr_with_hole(
+    ::Hole, 
+    cmn::ConstraintMatchVar, 
+    hole_location::Vector{Int}
+ )::Union{Tuple{Union{Int, Symbol}, Dict{Symbol, RuleNode}}, Nothing}
+    if hole_location == []
+        return cmn.var_name, Dict()
+    end
+    return nothing
+end
+
+# Matching RuleNode with ConstraintMatchNode
+"""
+Tries to match RuleNode `rn` with ConstraintMatchNode `cmn`.
+Returns a dictionary of values assigned to variables if the match is successful.
+"""
+function _match_expr(rn::RuleNode, cmn::ConstraintMatchNode)::Union{Dict{Symbol, RuleNode}, Nothing}
```

Comment:
this is a problem generally known as unification. Maybe we can leverage some insights from that are to compute this fast. Not a problem for now.

Reply:
Ah, I didn't know about that. Good to know!